File: modules/supabase-core-python/src/client/create_client.py
"""
Client Factory

Creates Supabase client instances with automatic environment detection
(local vs production).
"""

import os
from typing import Optional
from supabase import create_client as create_supabase_client, Client


def is_local_environment() -> bool:
    """Detect if running in local development environment."""
    return (
        os.getenv("ENVIRONMENT", "").lower() == "development"
        or os.getenv("NODE_ENV", "").lower() == "development"
        or os.getenv("DEBUG", "").lower() == "true"
        or os.getenv("SUPABASE_URL", "").startswith("http://localhost")
    )


def create_client(
    url: Optional[str] = None,
    anon_key: Optional[str] = None,
    service_role_key: Optional[str] = None,
) -> Client:
    """
    Creates a Supabase client instance for client-side or generic server-side use.
    Automatically detects local environment and uses local Supabase URL/keys if available.

    Args:
        url: Custom Supabase URL (overrides env var)
        anon_key: Custom anon key (overrides env var)
        service_role_key: Service role key (use with caution, bypasses RLS)

    Returns:
        Supabase client instance

    Example:
        ```python
        from supabase_core_python import create_client

        # Automatically detects local vs production environment
        supabase = create_client()

        # Use in your code
        response = supabase.table("posts").select("*").execute()
        ```
    """
    is_local = is_local_environment()

    supabase_url = url or (
        os.getenv("SUPABASE_URL_LOCAL") if is_local else os.getenv("SUPABASE_URL")
    )
    supabase_anon_key = anon_key or (
        os.getenv("SUPABASE_ANON_KEY_LOCAL") if is_local else os.getenv("SUPABASE_ANON_KEY")
    )
    supabase_service_role_key = service_role_key or (
        os.getenv("SUPABASE_SERVICE_ROLE_KEY_LOCAL")
        if is_local
        else os.getenv("SUPABASE_SERVICE_ROLE_KEY")
    )

    if not supabase_url or not supabase_anon_key:
        raise ValueError(
            "Missing Supabase configuration. Ensure SUPABASE_URL and SUPABASE_ANON_KEY "
            "are set in your environment variables."
        )

    # Use service role key if provided, otherwise use anon key
    key = supabase_service_role_key or supabase_anon_key

    return create_supabase_client(supabase_url, key)


def create_service_role_client(
    url: Optional[str] = None,
    service_role_key: Optional[str] = None,
) -> Client:
    """
    Creates a Supabase client instance with the service role key.
    This client bypasses Row Level Security (RLS) and should ONLY be used in
    secure server-side contexts (e.g., admin API routes, background jobs).

    WARNING: Never expose the service role key to client-side code!

    Args:
        url: Custom Supabase URL (overrides env var)
        service_role_key: Custom service role key (overrides env var)

    Returns:
        Supabase client instance with service role

    Example:
        ```python
        from supabase_core_python import create_service_role_client

        # Only use in secure server-side contexts
        admin_supabase = create_service_role_client()

        # Bypasses RLS - use with caution!
        response = admin_supabase.table("users").select("*").execute()
        ```
    """
    is_local = is_local_environment()
    supabase_url = url or (
        os.getenv("SUPABASE_URL_LOCAL") if is_local else os.getenv("SUPABASE_URL")
    )
    service_key = service_role_key or (
        os.getenv("SUPABASE_SERVICE_ROLE_KEY_LOCAL")
        if is_local
        else os.getenv("SUPABASE_SERVICE_ROLE_KEY")
    )

    if not supabase_url or not service_key:
        raise ValueError(
            "Missing Supabase configuration. Ensure SUPABASE_URL and "
            "SUPABASE_SERVICE_ROLE_KEY are set for service role client."
        )

    return create_supabase_client(supabase_url, service_key)

File: modules/supabase-core-typescript/src/client/create-client.ts
/**
 * Client Factory
 * 
 * Creates Supabase clients with automatic environment detection (local vs production).
 * Provides consistent client creation across the application.
 * 
 * Dependencies: @supabase/supabase-js
 */

import { createClient as createSupabaseClient, SupabaseClient } from '@supabase/supabase-js'
import type { Database } from '../../types/database-types'

/**
 * Environment detection: Determines if running in local development.
 * Checks for localhost URLs or explicit environment flag.
 */
function isLocalEnvironment(): boolean {
  const url = process.env.SUPABASE_URL || ''
  return url.includes('localhost') || url.includes('127.0.0.1') || process.env.NODE_ENV === 'development'
}

/**
 * Creates a Supabase client with automatic environment detection.
 * Uses environment variables to determine local vs production configuration.
 * 
 * @param options - Optional client configuration
 * @returns Configured Supabase client instance
 * 
 * @example
 * ```typescript
 * const supabase = createClient()
 * const { data } = await supabase.from('profiles').select('*')
 * ```
 */
export function createClient(options?: {
  /** Custom Supabase URL (overrides env var) */
  url?: string
  /** Custom anon key (overrides env var) */
  anonKey?: string
  /** Service role key (use with caution, bypasses RLS) */
  serviceRoleKey?: string
  /** Database types for type safety */
  dbTypes?: Database
}): SupabaseClient<Database> {
  const url = options?.url || process.env.SUPABASE_URL
  const key = options?.serviceRoleKey || options?.anonKey || process.env.SUPABASE_ANON_KEY

  if (!url || !key) {
    throw new Error(
      'Missing Supabase configuration. Ensure SUPABASE_URL and SUPABASE_ANON_KEY are set in environment variables.'
    )
  }

  const isLocal = isLocalEnvironment()

  return createSupabaseClient<Database>(url, key, {
    auth: {
      persistSession: !isLocal, // Don't persist in local dev (cookies handled by SSR)
      autoRefreshToken: true,
      detectSessionInUrl: true,
    },
    global: {
      headers: {
        'x-client-info': 'supabase-core-typescript@1.0.0',
      },
    },
  })
}

/**
 * Creates a Supabase client with service role key (bypasses RLS).
 * WARNING: Only use in server-side contexts (API routes, Edge Functions).
 * Never expose service role key to client-side code.
 * 
 * @param options - Optional client configuration
 * @returns Supabase client with service role permissions
 * 
 * @example
 * ```typescript
 * // In API route or Edge Function only
 * const adminSupabase = createServiceRoleClient()
 * const { data } = await adminSupabase.from('users').select('*') // Bypasses RLS
 * ```
 */
export function createServiceRoleClient(options?: {
  /** Custom Supabase URL (overrides env var) */
  url?: string
  /** Custom service role key (overrides env var) */
  serviceRoleKey?: string
  /** Database types for type safety */
  dbTypes?: Database
}): SupabaseClient<Database> {
  const url = options?.url || process.env.SUPABASE_URL
  const key = options?.serviceRoleKey || process.env.SUPABASE_SERVICE_ROLE_KEY

  if (!url || !key) {
    throw new Error(
      'Missing Supabase service role configuration. Ensure SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY are set.'
    )
  }

  return createSupabaseClient<Database>(url, key, {
    auth: {
      persistSession: false,
      autoRefreshToken: false,
    },
    global: {
      headers: {
        'x-client-info': 'supabase-core-typescript@1.0.0-service-role',
      },
    },
  })
}

